File: shop/src/main/java/basic/shop/member/command/domain/MemberId.java
package basic.shop.member.command.domain;

import lombok.AllArgsConstructor;
import lombok.Getter;

import javax.persistence.Access;
import javax.persistence.AccessType;
import javax.persistence.Column;
import javax.persistence.Embeddable;
import java.io.Serializable;
import java.util.Objects;

@Embeddable
@Getter
@AllArgsConstructor
public class MemberId implements Serializable {
    @Column(name = "member_id")
    private String id;

    @Override
    public boolean equals(Object o) {
        if(o == this)
            return true;
        if(!(o instanceof MemberId memberId))
            return false;
        return Objects.equals(id, memberId.id);
    }

    @Override
    public int hashCode() {
        return id == null ? 0 : id.hashCode();
    }
}

File: spring-mvc-with-kotlin/order-api/src/main/kotlin/basic/orderapi/member/command/domain/value/MemberId.kt
package basic.orderapi.member.command.domain.value

import java.io.Serializable
import javax.persistence.Column
import javax.persistence.Embeddable

@Embeddable
class MemberId(
    @Column(name = "member_id")
    var id: String? = null
): Serializable {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as MemberId

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    companion object {
        fun of(id: String): MemberId {
            return MemberId(id)
        }
    }
}